### PR TITLE
OEL-102: Structure and Service

### PR DESCRIPTION
**Pull request by saidatom**
### Description
In config/schema we will define variables: API key, database, and the endpoints
In config/install we will provide the defaults.

We will define a new service (yml and class).
The service class will have methods (described in it's interface) for handling the needed operations within the backend class, the method names will reflect the usage in the backend class methods.
The service class will be a wrapper around the ESC classes, making the needed API calls.

### Change log

- Added: EuropaSearchService, Config install and schema
- Changed: N/A
- Deprecated: N/A
- Removed: oe_search_demo, translations
- Fixed: N/A
- Security: N/A

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -12,3 +12,4 @@ vendor
 yarn.lock
 phpunit.xml
 docker-compose.*.yml
+rest
```
Comment (drishu) [inline .gitignore:15]: let's find a better setup

File: composer.json [hunk 2]
```diff
@@ -6,8 +6,16 @@
     "minimum-stability": "dev",
     "prefer-stable": true,
     "require": {
-        "drupal/core": "^8.7",
-        "php": ">=7.3"
+        "php": ">=7.3",
+        "drupal/core": "^8.7 || ^9.1",
+        "drupal/devel": "^4.1",
+        "drupal/facets": "^1.7",
+        "drupal/http_request_mock": "^1.0",
```
Comment (drishu) [inline composer.json:13]: move this to require-dev

File: composer.json [hunk 3]
```diff
@@ -6,8 +6,16 @@
...
+        "drupal/search_api": "^1.19",
+        "drupal/search_api_solr": "^3.3",
+        "http-interop/http-factory-guzzle": "^1.0",
```
Comment (drishu) [inline composer.json:16]: move this to require-dev

File: composer.json [hunk 4]
```diff
@@ -6,8 +6,16 @@
...
+        "openeuropa/europa-search-client": "dev-EPIC-EuropaSearch",
```
Comment (drishu) [inline composer.json:17]: let's use the develop branch for now to have all the client code avail

File: config/install/oe_search.server.yml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+api_key: europa_default
+database: EUROPA
+ingestion_api_endpoint: 'http://127.0.0.1:8080/'
```
Comment (drishu) [inline config/install/oe_search.server.yml:3]: http://web:8080

File: config/install/oe_search.server.yml [hunk 6]
```diff
@@ -0,0 +1,4 @@
...
+search_api_endpoint: 'http://127.0.0.1:8080/'
```
Comment (drishu) [inline config/install/oe_search.server.yml:4]: http://web:8080

Comment (drishu) [reply config/install/oe_search.server.yml:4]: Lets add also a token_api_endpoint

File: config/schema/oe_search.schema.yml [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+      label: 'API key'
+    database:
+      type: string
+      label: 'Database'
+    ingestion_api_endpoint:
+      type: string
+      label: 'Ingestion API endpoint URL'
+    search_api_endpoint:
+      type: string
+      label: 'Search API endpoint URL'
```
Comment (drishu) [inline config/schema/oe_search.schema.yml:16]: and a token_api_endpoint

File: oe_search.info.yml [hunk 8]
```diff
@@ -1,10 +1,10 @@
 name: OpenEuropa Search
 description: Search features for the OpenEuropa project.
 package: OpenEuropa
+dependencies:
+  - http_request_mock:http_request_mock
```
Comment (drishu) [inline oe_search.info.yml:5]: this is not really a dependency, everything concerning this should live in the test folder

File: src/EuropaSearchService.php [hunk 9]
```diff
@@ -0,0 +1,80 @@
+<?php
+
```
Comment (drishu) [inline src/EuropaSearchService.php:2]: declare(strict_types = 1);

File: src/EuropaSearchService.php [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+use Http\Adapter\Guzzle6\Client as HttpClient;
+use Http\Factory\Guzzle\RequestFactory;
+use Http\Factory\Guzzle\StreamFactory;
+use OpenEuropa\EuropaSearchClient\Api\IngestionApi;
+use OpenEuropa\EuropaSearchClient\Api\SearchApi;
+use OpenEuropa\EuropaSearchClient\Client;
+use OpenEuropa\EuropaSearchClient\ClientInterface;
+
+/**
+ * European Commission Europa Search.
```
Comment (drishu) [inline src/EuropaSearchService.php:15]:
```suggestion
 * Europa Search Service.
```

File: src/EuropaSearchService.php [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+  /**
+   * The config factory.
+   *
+   * @var \Drupal\Core\Config\ConfigFactoryInterface
+   */
+  protected $config;
+
+  /**
+   * Constructs a ProviderRepository instance.
```
Comment (drishu) [inline src/EuropaSearchService.php:27]:
```suggestion
   * Constructs a EuropaSearchService instance.
```

File: src/EuropaSearchService.php [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+   *   The config factory service.
+   */
+  public function __construct(
+    ConfigFactoryInterface $config_factory
+  ) {
+    $this->config = $config_factory->get('oe_search.server');
+  }
+
+  /**
+   * Return Ingestion API.
```
Comment (drishu) [inline src/EuropaSearchService.php:39]: add @return

File: src/EuropaSearchService.php [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+   */
+  public function ingestionApi() {
```
Comment (drishu) [inline src/EuropaSearchService.php:41]:
```suggestion
  public function ingestionApi(): IngestionApi {
```

File: src/EuropaSearchService.php [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+    $client = $this->getClient();
+    $api = new IngestionApi($client);
+
+    return $api;
+  }
+
+  /**
+   * Return Search results.
```
Comment (drishu) [inline src/EuropaSearchService.php:49]: add return

File: src/EuropaSearchService.php [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+   */
+  public function searchApi() {
```
Comment (drishu) [inline src/EuropaSearchService.php:51]:
```suggestion
  public function searchApi(): Search {
```

File: src/EuropaSearchService.php [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+   */
+  protected function getClient(): ClientInterface {
+    $configuration = $this->config->getRawData();
+
+    // Normalise configuration name from Drupal standards.
+    $configuration['apiKey'] = $configuration['api_key'];
+    unset($configuration['api_key']);
+    unset($configuration['_core']);
+
+    // @todo Make the client available through a service.
```
Comment (drishu) [inline src/EuropaSearchService.php:73]: we can remove this comment

File: src/Plugin/ServiceMock/ESIngestionPlugin.php [hunk 17]
```diff
@@ -0,0 +1,40 @@
+<?php
```
Comment (drishu) [inline src/Plugin/ServiceMock/ESIngestionPlugin.php:1]: let's move this to a module in test and make that dependent on http_request_mock

File: src/Plugin/ServiceMock/ESSearchPlugin.php [hunk 18]
```diff
@@ -0,0 +1,40 @@
+<?php
```
Comment (drishu) [inline src/Plugin/ServiceMock/ESSearchPlugin.php:1]: same, move it to the test module

File: tests/src/Kernel/ESCSearchMockTest.php [hunk 19]
```diff
@@ -0,0 +1,143 @@
+<?php
+
```
Comment (drishu) [inline tests/src/Kernel/ESCSearchMockTest.php:2]: declare(strict_types = 1);

File: tests/src/Kernel/ESCSearchMockTest.php [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  protected static $modules = ['oe_search', 'http_request_mock'];
+
+  /**
+   * Tests Europa Search Client Search.
+   */
+  public function testSearchApi(): void {
+    $document = new Document();
```
Comment (drishu) [inline tests/src/Kernel/ESCSearchMockTest.php:41]: Since we have the serializer here we can do something like:
`$expected_object = $this->getSerializer()->denormalize([array of values], Search::class);`

This way we also wont have to change, for example, the way setMetadata here, since denormalizer will take care of transforming it into an object if the signature changes.

File: tests/src/Kernel/ESCSearchMockTest.php [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+    $document->setSummary('string');
+    $document->setTitle('string');
+    $document->setUrl('string');
+    $document->setWeight(1.1);
+    $document->setChildren([]);
+
+    $documents = [$document];
+
+    $expected = new Search();
+    $expected->setResults($documents);
```
Comment (drishu) [inline tests/src/Kernel/ESCSearchMockTest.php:66]: we could actually create a bigger array that represents the whole search object with results and all and call denormalize on that.

File: tests/src/Kernel/ESCSearchMockTest.php [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    // Mock EuropaSearchService service method searchApi
+    $serviceMock = $this->getMockBuilder(EuropaSearchService::class)
+      ->disableOriginalConstructor()
+      ->getMock();
+
+    $serviceMock->expects($this->any())
+      ->method('searchApi')
+      ->will($this->returnValue($actual_docs));
```
Comment (drishu) [inline tests/src/Kernel/ESCSearchMockTest.php:88]: if we do it like this we actually test the mocked servers response, not the service method, same for testIngestionApi

File: .gitattributes [hunk 23]
```diff
@@ -0,0 +1,61 @@
+# Drupal git normalization
```
Comment (drishu) [inline .gitattributes:1]: is this file really needed ? I dont see it in any of the other openeuropa repos

**Review by drishu — CHANGES_REQUESTED** [review]
I will finish my review here for now as these changes need to be implemented before I can continue.

Also please look at the messages in drone.

File: .gitattributes (hunk 23 above)
Comment (claudiu-cristea) [reply .gitattributes:1]: This is already provided on Drupal install level. We don't need it.

File: composer.json [hunk 24]
```diff
@@ -6,8 +6,16 @@
     "minimum-stability": "dev",
     "prefer-stable": true,
     "require": {
-        "drupal/core": "^8.7",
-        "php": ">=7.3"
+        "php": ">=7.3",
+        "drupal/core": "^8.7 || ^9.1",
```
Comment (claudiu-cristea) [inline composer.json:10]: 8.7 is not supported anymore. Nor 8.8. Drupal 8.9, as a LTS, [is supported until November this year](https://hugovk.github.io/drupal-release-cycle/). So, maybe, we should only use the major versions: `^8 || ^9`

File: composer.json [hunk 25]
```diff
@@ -6,8 +6,16 @@
...
+        "drupal/devel": "^4.1",
+        "drupal/facets": "^1.7",
```
Comment (claudiu-cristea) [inline composer.json:12]: I don't think `drupal/facets` should be a _hard_ dependency. There might be projects that only need simple search but no facets. But for sure we want the test facets integration so this should go in `require-dev`.

File: composer.json [hunk 26]
```diff
@@ -6,8 +6,16 @@
     "minimum-stability": "dev",
     "prefer-stable": true,
     "require": {
-        "drupal/core": "^8.7",
-        "php": ">=7.3"
+        "php": ">=7.3",
+        "drupal/core": "^8.7 || ^9.1",
+        "drupal/devel": "^4.1",
```
Comment (claudiu-cristea) [inline composer.json:11]: This is only used for local development. Is not used in production neither in tests. I think it can be removed or moved as dev dependency.

File: composer.json [hunk 27]
```diff
@@ -6,8 +6,16 @@
...
+        "drupal/facets": "^1.7",
+        "drupal/http_request_mock": "^1.0",
+        "drupal/search_api": "^1.19",
+        "drupal/search_api_solr": "^3.3",
```
Comment (claudiu-cristea) [inline composer.json:15]: Really? How this module relies on _Search API Solr_? Or I'm missing something?

File: config/install/oe_search.server.yml [hunk 28]
```diff
@@ -0,0 +1,4 @@
+api_key: europa_default
```
Comment (claudiu-cristea) [inline config/install/oe_search.server.yml:1]: We should not store connection settings outside the server config entity. That is `search_api.server.*`.

File: config/install/oe_search.server.yml (hunk 6 above)
Comment (claudiu-cristea) [reply config/install/oe_search.server.yml:4]: > Lets add also a token_api_endpoint

~~No, that's sensitive data and should live in `settings.php` as a setting. Otherwise, some project might push that in VCS as config sync YAMLs.~~ EDIT: Sorry, I wrongly thought that it's about a key/secret.

File: config/schema/oe_search.schema.yml (hunk 7 above)
Comment (claudiu-cristea) [reply config/schema/oe_search.schema.yml:16]: ~~Same. Sensitive info should not end in config.~~ EDIT: Sorry, I wrongly thought that it's about a key/secret.

File: oe_search.info.yml (hunk 8 above)
Comment (claudiu-cristea) [reply oe_search.info.yml:5]: True, not a dependency. But we can add it as `test_dependencies`

File: oe_search.info.yml [hunk 29]
```diff
@@ -1,10 +1,10 @@
...
+  - search_api_solr:search_api_solr
```
Comment (claudiu-cristea) [inline oe_search.info.yml:6]: For sure _Search API Solr_ is not a dependency

File: oe_search.info.yml [hunk 30]
```diff
@@ -1,10 +1,10 @@
...
+  - facets:facets
```
Comment (claudiu-cristea) [inline oe_search.info.yml:7]: Same as for `composer.json`, this is a dev dependency. Move it under `test_dependencies:`.

File: oe_search.info.yml [hunk 31]
```diff
@@ -1,10 +1,10 @@
...
 
 type: module
-core: 8.x
-
-'interface translation project': oe_search
-# The path to the actual translations is defined in
-# oe_search_locale_translation_projects_alter().
+core_version_requirement: ^8.8 || ^9
```
Comment (claudiu-cristea) [inline oe_search.info.yml:10]: Same as in `composer.json`, `^8 || ^9`

File: oe_search.services.yml [hunk 32]
```diff
@@ -0,0 +1,4 @@
+services:
+  oe_search.es_service:
```
Comment (claudiu-cristea) [inline oe_search.services.yml:2]: I've been discussed with @drishu, in chat, we will not provide a service wrapper. We'll use the Europa Search Client directly. And we'll open a follow-up to wrap it into a plugin, more or less on the same way _Search API Solr_ is doing with the connector plugins.

File: src/EuropaSearchService.php [hunk 33]
```diff
@@ -0,0 +1,80 @@
...
+use Http\Factory\Guzzle\StreamFactory;
+use OpenEuropa\EuropaSearchClient\Api\IngestionApi;
+use OpenEuropa\EuropaSearchClient\Api\SearchApi;
+use OpenEuropa\EuropaSearchClient\Client;
+use OpenEuropa\EuropaSearchClient\ClientInterface;
+
+/**
+ * European Commission Europa Search.
+ */
+class EuropaSearchService {
```
Comment (claudiu-cristea) [inline src/EuropaSearchService.php:17]: I'm gonna ignore this file, as we've discussed to drop this wrapping service.

File: src/Plugin/ServiceMock/ESIngestionPlugin.php (hunk 17 above)
Comment (claudiu-cristea) [reply src/Plugin/ServiceMock/ESIngestionPlugin.php:1]: True, this is only used for testing.

File: src/Plugin/ServiceMock/ESIngestionPlugin.php [hunk 34]
```diff
@@ -0,0 +1,40 @@
...
+   */
+  public function applies(RequestInterface $request, array $options): bool {
+    return $request->getUri()->getHost() === 'api.com' && $request->getUri()->getPath() === '/ingestion';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getResponse(RequestInterface $request, array $options): ResponseInterface {
+    $response = '{"apiVersion": "string","reference": "string","trackingId": "string"}';
```
Comment (claudiu-cristea) [inline src/Plugin/ServiceMock/ESIngestionPlugin.php:35]: Just for developer readability, I'd prefer that we define a PHP array and then serialize it with the JSON serializer.

**Review by claudiu-cristea — CHANGES_REQUESTED** [review]
I had a call with @drishu today on OEL-101, OEL-102. There are several issues with the initial approach:

1. We should not store the connection info into an arbitrary config entity such as `oe_search.server`. Search API already offers the `search_api_server` config entities (`search_api.server.*`) and they are the standard way to define a server/backend, including its connection info. Using a custom config has some drawbacks:
   * Not a standard _Search API_ approach.
   * It allows defining ONLY a single server/backend.
1. We will not wrap the _Europa Search Client_ in a service. True, this was the initial analysis but now, reading the code, it turns out to be wrong. Looking at _Search API Solr_ module, it seems that we can learn something from their approach. They are defining a connector plugin type to wrap the PHP client. However, it's very unlikely that we'll hit a use-case where a project would use a different client than `openeuropa/europa-search-client` in the next 10 years. So, we've decided to:
   * Instantiate the client directly in our _Search API_ server plugin.
   * Open a follow-up to implement a _Europa Search Client_ plugin type. The _Search API_ server plugin will have a new configuration to instruct which client plugin to use. This module will ship a default plugin that would provide the `openeuropa/europa-search-client` as client.

Give all these decisions/changes, I'm working now to re-scope the tickets.

File: config/schema/oe_search.schema.yml (hunk 7 above)
Comment (claudiu-cristea) [reply config/schema/oe_search.schema.yml:16]: Sorry. I was wrong

**Comment by claudiu-cristea** [comment]
As per https://citnet.tech.ec.europa.eu/CITnet/jira/browse/OEL-102?focusedCommentId=4712258&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-4712258, I'll merge this PR
